HF component update to not use fsspec components

ankitageorge opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #159681
* #159406
* __->__ #159405

Update HF components to not inherit from fsspec components and instead use filesystem writer/reader. The reason is because there doesn't seem to be much of a need for fsspec, since users are using mounted storage. Using local storage will allow for performance improvements because we can take advantage of the safe_open API provided by HF safetensors (30s vs 4s for load of 8b model), which is signifcant performance wins over reading bytes and converting to tensors which is what we are doing now. Also, we can use the official methods provided by HF instead of relying on reading the metadata by bytes and loading it

Differential Revision: [D78993550](https://our.internmc.facebook.com/intern/diff/D78993550/)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

ankitageorge committed:
HF component update to not use fsspec components

Update HF components to not inherit from fsspec components and instead use filesystem writer/reader. The reason is because there doesn't seem to be much of a need for fsspec, since users are using mounted storage. Using local storage will allow for performance improvements because we can take advantage of the safe_open API provided by HF safetensors, which is signifcant performance wins over reading bytes and converting to tensors which is what we are doing now.

Differential Revision: [D78993550](https://our.internmc.facebook.com/intern/diff/D78993550/)

[ghstack-poisoned]

ankitageorge committed:
Update on "HF component update to not use fsspec components"


Update HF components to not inherit from fsspec components and instead use filesystem writer/reader. The reason is because there doesn't seem to be much of a need for fsspec, since users are using mounted storage. Using local storage will allow for performance improvements because we can take advantage of the safe_open API provided by HF safetensors, which is signifcant performance wins over reading bytes and converting to tensors which is what we are doing now.

Differential Revision: [D78993550](https://our.internmc.facebook.com/intern/diff/D78993550/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

ankitageorge committed:
Update on "HF component update to not use fsspec components"


Update HF components to not inherit from fsspec components and instead use filesystem writer/reader. The reason is because there doesn't seem to be much of a need for fsspec, since users are using mounted storage. Using local storage will allow for performance improvements because we can take advantage of the safe_open API provided by HF safetensors, which is signifcant performance wins over reading bytes and converting to tensors which is what we are doing now.

Differential Revision: [D78993550](https://our.internmc.facebook.com/intern/diff/D78993550/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

saumishr commented on torch/distributed/checkpoint/hf_storage.py:
```diff
@@ -52,7 +52,7 @@
 
 
-class HuggingFaceStorageWriter(FsspecWriter):
+class HuggingFaceStorageWriter(FileSystemWriter):
```
@ankitageorge If any user needs the FSSpec based component, they should be able to write a wrapper extending the HuggingFaceStorageWriter. Could you please confirm if thats the case.

saumishr commented on torch/distributed/checkpoint/hf_storage.py:
```diff
@@ -52,7 +52,7 @@
 
 
-class HuggingFaceStorageWriter(FsspecWriter):
+class HuggingFaceStorageWriter(FileSystemWriter):
```
Will this be backward compatible? Its a very new feature so probably no major adoption yet but we enabled a few partners to start using the FSSpec component.

ankitageorge replied:
The only usage I could find is torchtune, but considering it's being deprecated I think it is okay. The flow will just go back to users having to download files from HF hub before running, which was the old experience.

ankitageorge committed:
Update on "HF component update to not use fsspec components"


Update HF components to not inherit from fsspec components and instead use filesystem writer/reader. The reason is because there doesn't seem to be much of a need for fsspec, since users are using mounted storage. Using local storage will allow for performance improvements because we can take advantage of the safe_open API provided by HF safetensors (30s vs 4s for load of 8b model), which is signifcant performance wins over reading bytes and converting to tensors which is what we are doing now. Also, we can use the official methods provided by HF instead of relying on reading the metadata by bytes and loading it

Differential Revision: [D78993550](https://our.internmc.facebook.com/intern/diff/D78993550/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
